Add a lint rule for e-mail

OnkarRuikar opened the pull request:
With this the linter will not allow `e-mail`.

Josh-Cena commented on .markdownlint-cli2.jsonc:
```diff
@@ -189,6 +189,13 @@
           "skipCode": true
+        },
+        {
+          "name": "spelling-email",
+          "message": "Use 'email'",
+          "search": "e-mail",
+          "replace": "email",
+          "skipCode": false
```
I suggest making it a generic rule like "bad-spelling" so we don't have to add 7 lines to this file every time something gets banned.

Also, does cSpell have a "block-list" sort of feature?

OnkarRuikar replied:
Implemented https://github.com/mdn/content/pull/23146#issuecomment-1367795969

OnkarRuikar commented:
> I suggest making it a generic rule like "bad-spelling" so we don't have to add 7 lines to this file every time something gets banned.
> 
> Also, does cSpell have a "block-list" sort of feature?

@Josh-Cena the solution depends on how far we are going to go with spellings:
a. occasionally block some outdated words like `e-mail`, `web site` etc.
b. have [full-blown list of correct words](https://developers.google.com/style/word-list#at-scale) like Microsoft and Google docs and enforce them.
c. have dedicated spellchecker for everything and all words.

For option #a we can keep adding words like this PR does.

For option #b we implement dedicated replacement list rule in the linter:
```
{
	name: "bad-spellings"
    search: [ "e-mail", "web site", "time stamp"],
    replace: ["email", "website", "timestamp"]
}
```


For option #c we can use tools like `cspell` and `spellchecker-cli`.
The cspell does have [forbid words](https://cspell.org/docs/forbidden-words/) feature.
```bash
> cspell --show-suggestions index.md

.../index.md:2:8 - Unknown word (heiallo) Suggestions: [hello, hallo, niello, heirloom, heimdall]
.../index.md:13:1 - Forbidden word (e-mail) Suggestions: [email, emails, entail, emil, camail]
CSpell: Files checked: 1, Issues found: 2 in 1 files
```
I think for the long-run we can use cSpell, as cspell will integrate well with VSCode as well because it's port of VSCode plugin "Code Spell Checker".

Ping @bsmth, @nschonni

Josh-Cena commented:
I know @nschonni is already running cSpell locally, and I also have a copy of his config for my own VS Code config, so if he's taking care of other incorrect spellings anyway, this should be handleable as well.

bsmth commented:
Thanks @OnkarRuikar. I agree with Josh that this is a lot of work per disallowed word and we should look at alternatives. Cspell might be a good avenue, I'm also using it in VSCode and adding words to a user config as I go, but it's still going to need some tweaks. If you run it over `web/javascript/**/*.md` with a config that already has some hand-picked additions:

```bash
cspell -c ~/mdncspell.json --unique --no-progress files/en-us/web/javascript/**/*.md
...
./web/javascript/reference/statements/do...while/index.md:19:47 - Unknown word (dowhile)
./web/javascript/reference/statements/export/index.md:270:57 - Unknown word (htmlelement)
./web/javascript/reference/statements/for-await...of/index.md:21:47 - Unknown word (forawaitof)
./web/javascript/reference/statements/for...in/index.md:16:47 - Unknown word (forin)
./web/javascript/reference/statements/for...of/index.md:19:47 - Unknown word (forof)
./web/javascript/reference/statements/function_star_/index.md:290:13 - Unknown word (Lindesay)
./web/javascript/reference/statements/import/index.md:229:3 - Unknown word (Limin)
./web/javascript/reference/statements/import/index.md:229:20 - Unknown word (Terlson)
./web/javascript/reference/statements/let/index.md:98:5 - Unknown word (Redeclarations)
./web/javascript/reference/statements/throw/index.md:134:7 - Unknown word (ZIPCODE)
./web/javascript/reference/statements/throw/index.md:158:17 - Unknown word (rethrown)
./web/javascript/reference/statements/try...catch/index.md:17:47 - Unknown word (trycatch)
./web/javascript/reference/statements/try...catch/index.md:100:3 - Unknown word (myroutine)
./web/javascript/reference/statements/var/index.md:23:5 - Unknown word (varname)
./web/javascript/reference/strict_mode/index.md:120:39 - Unknown word (Varible)
./web/javascript/reference/strict_mode/index.md:164:109 - Unknown word (undeletable)
./web/javascript/reference/template_literals/index.md:126:12 - Unknown word (collapser)
./web/javascript/reference/template_literals/index.md:331:72 - Unknown word (DSLs)

CSpell: Files checked: 904, Issues found: 1964 in 471 files
```

Approx 2000 unknown words under JS docs, so that rules out option a) at least for all types of typos (including uncommon words). I like the links you have pointed to for word lists, these would be useful to include in a spellchecker for common typos (style enforcement).

I think the markdownlint job should just check formatting and we can tackle spelling in a separate effort that's easier to maintain, what do you think?

OnkarRuikar commented:
From the conversation in `#web-docs-content` room, involving full fledged spellchecker for technical documentation like this doesn't look like a good idea.
> **Onkar:** All, we recently corrected the e-mail word in the entire content repo.
Adding to that, can we implement a full-blown spell checker for mdn/content repo?
It'll work in local environment(VSCode and cli), husky, pr-checks, and nightly cron jobs.
For more info refer: https://github.com/mdn/content/pull/23146#issuecomment-1363645280
>
> **@Rumyra:** IIRC When we moved to github a couple of years back we had someone run a spell checker over the entire repo and it was too much, changing the meaning of things, so we don't recommend it. However reading the thread I'm not opposed to option a or b 👍
>
> **@schalkneethling:** Yeah, spell checkers that works well with technical documentation and code still has some way to go. I wonder if the latest things from OpenAI and HuggingFace might help here?

I agree with the argument because we'll have to add a lot of words to our custom allow list like: abortable, componentized, accentcolortext, appcache, redirectors, authorid, mergeable  etc.
Source of new words could be:
- variable names: e.g. msgtext. The spellcheckers do detect camelCase but our code is not completely camel cased.
- [lorem ipsum](https://www.lipsum.com/): Almost all the words in such text are invalid. :scream: 
- convenience:  e.g. abortable, mergeable.

Even if we add these to our ignore list and setup a spellchecker, we'll have to keep updating the list for new contributions or ask PR authors to correct them.

Middle way is to maintain a bad word list ([like this](https://developers.google.com/style/word-list#at-scale)) only for words which we don't want to promote e.g. e-mail, web site. I am working on option b. It'll add ability to specify arrays in the rule config in `markdownlint-rule-search-replace` package.

***
@bsmth 
> I think the markdownlint job should just check formatting and we can tackle spelling in a separate effort that's easier to maintain, what do you think?

Difference between handling it in linter vs spellchecker would be ability to auto fix. The linter will be able to autofix `e-mail` to `email`. Spellchecker won't be able to do that, it'll only flag the bad word.
If we don't suggest corrections then contributors will be forced to find replacements on their own.

If we want to to add a spellchecker then bigger discussion is needed. May be on https://github.com/mdn/mdn-community/discussions?

OnkarRuikar committed:
Add a generic bad-spelling rule

OnkarRuikar commented:
All, I've implemented and added a new generic rule for handling the incorrect spellings. Now the config will like this:
```
        {
          "name": "bad-spelling",
          "message": "Incorrect spelling",
          "searchPattern": ["/e-mail/ig", 
                            "/time stamp/ig", 
                            "/web site|web-site/ig"],

          "replace": ["email", "timestamp", "website"],
          "skipCode": false
        }
```
To add to above discussions, using linter we can ban split words like `web site`. Spell checkers will always see them as valid.

Ping @nschonni.

schalkneethling commented on .markdownlint-cli2.jsonc:
```diff
@@ -189,6 +189,13 @@
           "skipCode": true
+        },
+        {
+          "name": "bad-spelling",
```
nit:
```suggestion
          "name": "incorrect-spelling",
```

Not sure if the value of the `name` property is defined by the library, but if not, I would recommend renaming it to something a little less bad 😄

OnkarRuikar replied:
No rule has been hard coded in [the library](https://github.com/OnkarRuikar/markdownlint-rule-search-replace/#configuration).

We can create more rules like this:
```
{
  "default": true,
  "search-replace": {
    "rules": [
      {
        "name": "product-names",
        "message": "Use correct cases for trademarks and registered names",
        "search": ["OS x", "Javascript", "Github", "Typescript"],
        "replace": ["macOS", "JavaScript", "GitHub", "TypeScript"],
        "skipCode": false
      }
    ]
  }
}
```

schalkneethling reviewed (commented):
Small suggestions that might not be possible. Thanks, @OnkarRuikar

OnkarRuikar committed:
Apply suggestions from code review

Co-authored-by: Schalk Neethling <schalk.neethling@mechanical-ink.co.za>

bsmth commented:
Thank @OnkarRuikar. I was initially thinking it should be "either or" with the markdownlint job versus a dedicated spellchecker, but maybe we could make use of both. It looks like checking in a cspell config _at some point_ might be a good idea regardless of how it's enforced (I spotted you can include a lorem-ipsum dictionary, too, see [this gist](https://gist.github.com/bsmth/dbac540fafb85b61a50ae381e9dd29c0)). Warnings via spellcheck on changed files in PRs combined with adding flagged words to the config as we go might help.

>Difference between handling it in linter vs spellchecker would be ability to auto fix.

Yeah this is useful.

>I think for the long-run we can use cSpell, as cspell will integrate well with VSCode as well

I agree with this. I think this PR is a useful change for the meantime 👍🏻 

Some notes / things to look at alongside this PR or for a separate spellchecker discussion:

- [Action to generate cspell warnings on PRs](https://github.com/streetsidesoftware/cspell-action)
- [Azure JS SDK docs for spellcheck in CI](https://github.com/Azure/azure-sdk-tools/blob/main/doc/common/spellcheck.md#spell-check-in-ci) and the [cspell config](https://github.com/Azure/azure-sdk-for-js/blob/main/.vscode/cspell.json)

OnkarRuikar commented:
One more point to add to the discussion, what to do with incorrect spellings coming from external resources?
For example,
````md
a. In URLs:
[Colour Manegement Guide](www.blogs.com/colour-manegement-guide)

b. Article titles:
For more info refer the article "Using CSS Property appearence" on www.abc.com.

c. In external code snippets:
Consider following code snippet taken from the example ( www.github.com/mdn/interactive-examples/abcd.html)
```js
let colar = 'green';
```
````
Should we correct it on our side? Or Do we add a setting/magic comment to ignore the spot?

caugner reviewed (approved):
Config change LGTM.

OnkarRuikar commented:
> Preview URLs (294 pages)

@yin1999 and @caugner Looks like there is a bug in preview companion. Only one file was modified but it generated preview for 294 files.

yin1999 commented:
> > Preview URLs (294 pages)
> 
> @yin1999 and @caugner Looks like there is a bug in preview companion. Only one file was modified but it generated preview for 294 files.

It looks like the `head` and `base` hash is not right in this PR, comparing between [github API](https://api.github.com/repos/mdn/content/pulls/23146) and [the github workflow run](https://github.com/mdn/content/actions/runs/4221399464/jobs/7328643877#step:5:4).

OnkarRuikar commented:
> It looks like the `head` and `base` hash is not right in this PR, comparing between [github API](https://api.github.com/repos/mdn/content/pulls/23146) and [the github workflow run](https://github.com/mdn/content/actions/runs/4221399464/jobs/7328643877#step:5:4)

Seems like GitHub picked up the detached commit https://github.com/mdn/content/commit/ca3b60c0f9b7d9379c6d560b1ef5d76705a0f2d1 as base commit, then workflow started working on it then then GitHub course corrected itself. Will putting some delay in workflow before fetching changed files solve the issue? It'll be hard to reproduce and test though. :(
